src/algorithms.jl

```diff
@@ -215,7 +257,53 @@ macro functiondef(f)
         # copy documentation to both functions
         Core.@__doc__ $f, $f!
     end
-    return esc(ex)
+    ex_double_arg = quote
+        # out of place to inplace
+        $f(A, B; kwargs...) = $f!(copy_input($f, A, B)...; kwargs...)
+        $f(A, B, alg::AbstractAlgorithm) = $f!(copy_input($f, A, B)..., alg)
+
+        # fill in arguments
+        function $f!(A, B; alg=nothing, kwargs...)
+            return $f!(A, B, select_algorithm($f!, (A, B), alg; kwargs...))
+        end
+        function $f!(A, B, out; alg=nothing, kwargs...)
+            return $f!(A, B, out, select_algorithm($f!, (A, B), alg; kwargs...))
+        end
+        function $f!(A, B, alg::AbstractAlgorithm)
+            return $f!(A, B, initialize_output($f!, A, B, alg), alg)
+        end
+
+        # define fallbacks for algorithm selection
+        @inline function select_algorithm(::typeof($f), A, alg::Alg; kwargs...) where {Alg}
+            return select_algorithm($f!, A, alg; kwargs...)
+        end
+        # define default algorithm fallbacks for out-of-place functions
+        # in terms of the corresponding in-place function
+        @inline function default_algorithm(::typeof($f), A, B; kwargs...)
+            return default_algorithm($f!, A, B; kwargs...)
+        end
+        # define default algorithm fallbacks for out-of-place functions
+        # in terms of the corresponding in-place function for types,
+        # in principle this is covered by the definition above but
+        # it is necessary to avoid ambiguity errors with the generic definitions:
+        # ```julia
+        # default_algorithm(f::F, A; kwargs...) where {F} = default_algorithm(f, typeof(A); kwargs...)
+        # function default_algorithm(f::F, ::Type{T}; kwargs...) where {F,T}
+        #     throw(MethodError(default_algorithm, (f, T)))
+        # end
+        # ```
+        @inline function default_algorithm(::typeof($f), ::Type{A}, ::Type{B}; kwargs...) where {A, B}
+            return default_algorithm($f!, A, B; kwargs...)
+        end
+
+        # copy documentation to both functions
+        Core.@__doc__ $f, $f!
+    end
+    if f_n_args == 1
+        return esc(ex_single_arg)
+    elseif f_n_args == 2
+        return esc(ex_double_arg)
+    end
```

Comment:
It might be cleaner to put this if-clause before the actual expressions are constructed, and only construct a single one?

Reply:
definitely agreed!

Reply:
Done! Used a `Val` so we can dispatch nicely and would work with switching to dispatching from an `NTuple` in the future.